The Genome Aggregation Database (gnomAD), is a coalition of investigators seeking to aggregate and harmonize exome and genome sequencing data from a variety of large-scale sequencing projects, and to make summary data available for the wider scientific community. The project is overseen by co-directors Heidi Rehm and Mark Daly, and steering committee members Daniel MacArthur, Benjamin Neale, Michael Talkowski, Anne O'Donnell-Luria, Konrad Karczewski, Grace Tiao, Matthew Solomonson, and Kat Tarasova. In its first release, which contained exclusively exome data, it was known as the Exome Aggregation Consortium (ExAC).

The v3.1 short variant data set provided on this website spans 76,156 genomes from unrelated individuals sequenced as part of various disease-specific and population genetic studies, and is aligned against the GRCh38 reference. See the [gnomAD v3.1 blog post](https://gnomad.broadinstitute.org/blog/2020-10-gnomad-v3-1-new-content-methods-annotations-and-data-availability/) for details of the latest release. The v2 short variant data set provided on this website spans 125,748 exomes and 15,708 genomes from unrelated individuals sequenced as part of various disease-specific and population genetic studies, totalling 141,456 individuals, and is aligned against the GRCh37 reference. See the [gnomAD blog](https://gnomad.broadinstitute.org/blog/) for details of previous releases ([v3.0](https://gnomad.broadinstitute.org/blog/2019-10-gnomad-v3-0/), [v2.1](https://gnomad.broadinstitute.org/blog/2018-10-gnomad-v2-1/)). We have removed individuals known to be affected by severe pediatric disease, as well as their first-degree relatives, so these data sets should serve as useful reference sets of allele frequencies for severe pediatric disease studies - however, note that some individuals with severe disease may still be included in the data sets, albeit likely at a frequency equivalent to or lower than that seen in the general population.

All of the raw data from these projects have been reprocessed through equivalent pipelines, and jointly variant-called to increase consistency across projects. The processing pipelines were written in the [WDL workflow definition language](https://software.broadinstitute.org/wdl/) and executed using the [Cromwell execution engine](https://github.com/broadinstitute/cromwell), open-source projects for defining and executing genomic workflows at massive scale on multiple platforms. The gnomAD data set contains individuals sequenced using multiple exome capture methods and sequencing chemistries, so coverage varies between individuals and across sites. This variation in coverage is incorporated into the variant frequency calculations for each variant.

gnomAD was QCed and analysed using the [Hail](https://hail.is/) open-source framework for scalable genetic analysis.

A list of gnomAD Principal Investigators and groups that have contributed data and analysis to the current release is available below.

The generation of this call set was funded primarily by the Broad Institute, and the data here are released publicly for the benefit of the wider biomedical community. There are no publication restrictions or embargoes on these data. Please [cite the flagship gnomAD paper](https://www.nature.com/articles/s41586-020-2308-7#citeas) for any use of these data.

The gnomAD structural variant (SV) v2.1 data set provided on this website spans 10,847 genomes from unrelated individuals sequenced as part of various disease-specific and population genetic studies, and is aligned against the GRCh37 reference. It mostly (but not entirely) overlaps with the genome set used for the gnomAD short variant release. See the [Structural variants in gnomAD blog post](https://gnomad.broadinstitute.org/blog/2019-03-structural-variants-in-gnomad/) for details of the SV release. As with the gnomAD short variant data set, we have removed individuals known to be affected by severe pediatric disease, as well as their first-degree relatives, so this data set should serve as a useful reference set of allele frequencies for severe pediatric disease studies - however, note that some individuals with severe disease may still be included in the data set, albeit likely at a frequency equivalent to or lower than that seen in the general population.

All of the raw data from these projects have been reprocessed through the same pipeline, and jointly variant-called to increase consistency across projects.In brief, we aggregated Illumina short-read WGS from population genetic and complex disease-association studies. We discovered SVs by integrating four published SV algorithms ([Manta](https://www.ncbi.nlm.nih.gov/pubmed/26647377), [DELLY](https://www.ncbi.nlm.nih.gov/pubmed/22962449), [MELT](https://www.ncbi.nlm.nih.gov/pubmed/28855259), and [cn.MOPS](https://www.ncbi.nlm.nih.gov/pubmed/22302147)) to identify SVs across seven mutational classes, and jointly filtered, genotyped, resolved, and annotated these SVs across all genomes. All SV discovery was performed in [FireCloud](https://software.broadinstitute.org/firecloud/) using the [Workflow Description Language (WDL)](https://software.broadinstitute.org/wdl/) and the [Cromwell Execution Engine](https://cromwell.readthedocs.io/en/stable/), where the components of the gnomAD SV discovery pipeline are available as public methods with dedicated Docker images. Extensive technical details of this process are provided in the supplementary information of the [gnomAD SV paper](https://broad.io/gnomad_sv); please cite this paper for use of the SV data.

The data released by gnomAD are available free of restrictions under the [Creative Commons Zero Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/). This means that you can use it for any purpose without legally having to give attribution. However, we request that you actively acknowledge and give attribution to the gnomAD project, and link back to the relevant page, wherever possible. Attribution supports future efforts to release other data. It also reduces the amount of "orphaned data", helping retain links to authoritative sources.

The aggregation and release of summary data from the exomes and genomes collected by the Genome Aggregation Database has been approved by the Partners IRB (protocol 2013P001339, "Large-scale aggregation of human genomic data").
