---
question: "Why doesn't the amino acid position and consequence on the gene page match what I expect for my transcript and how does gnomAD choose which transcript to use for annotating variants?"
---

The gene page summarizes the most severe functional consequence and protein-level change regardless of the transcript. gnomAD v4 uses the MANE Select transcript as the primary transcript for annotating HGVS nomenclature/consequence. MANE stands for Matched Annotation from NCBI and EMBL-EBI and MANE Select identifies one high-quality representative transcript per protein-coding gene that is well-supported by experimental data and represents the biology of the gene. More background on the MANE project can be found [here](https://www.ncbi.nlm.nih.gov/refseq/MANE/). If no MANE Select transcript exists, the Ensembl canonical transcript is selected. A "â€ " denotes a consequence that is for a non-MANE Select transcript (or non-canonical transcript if no MANE Select exists). On the gene page, you can select which transcript you would like to view. The MANE Select transcript (or canonical if no MANE Select exists) is marked by "\*". Please note that v2 still uses the canonical transcript because MANE Select transcripts are derived from GRCh38. For information on the VEP and GENCODE annotation versions used please see our [annotation section](/help/vep). Please note, the browser represents transcript versions used at the time of release, which may differ from more recent versions selected by the MANE Select project. Please check the version number listed by the "MANE Select transcript" attribute at the top of the gene summary page and check if it is the same or different from the version listed in the browser's "Show Transcript" list to be warned of possible discrepancies.
