---
id: str-read-data-multiple-repeat-units
title: 'Read Visualizations'
---

Read visualizations were generated using [REViewer](https://www.illumina.com/science/genomics-research/articles/reviewer-alignments-short-reads-long-repeat.html). They are presented here in order from longest to shortest allele size. For additional information and examples on how to interpret these images, see the [gnomAD STR blog post](https://gnomad.broadinstitute.org/news/2022-01-the-addition-of-short-tandem-repeat-calls-to-gnomad/) as well as [Dolzhenko 2022](https://doi.org/10.1101/2021.10.20.465046).

Since different motifs have been observed at this locus, we ran the [call_non_ref_pathogenic_motifs.py](https://github.com/broadinstitute/str-analysis/blob/main/str_analysis/call_non_ref_pathogenic_motifs.py) script on each sample to first detect the motif(s) present at this locus and then run ExpansionHunter and REViewer on those specific motif(s). Since these loci are autosomal, each individual inherits two copies of this STR locus, and these 2 copies may contain different motifs. When the call_non_ref_pathogenic_motifs.py script detects only one motif in an individual, it simply runs ExpansionHunter and REViewer for that motif. However, if the script detects two different motifs, it runs ExpansionHunter and REViewer twice - once for each motif. This yields two separate genotypes, and two different REViewer images. To combine the two REViewer images, the script simply concatenates them into a single image after adding some title text and a thick horizontal line separating the two. This combined image allows users to examine the read data for either motif. Then, to collapse the two genotypes into a single diploid genotype, the script selects a short allele size from one genotype, and a long allele size from the other genotype. These collapsed genotypes have low accuracy and are mainly reported for completeness. We recommend reviewing read visualizations to evaluate any genotypes of interest.
